statsmodels/regression/mixed_glm.py

```diff
@@ -0,0 +1,1514 @@
+
+        Parameters
+        ----------
+        params : array-like, 1d
+            The fixed effects parameters
+        cov_re : array-like, 2d
+            The covariance matrix of the random effects
+        scale : float
+            The overdispersion scale parameter for the exponential family.
+        omethod : string
+            The optimization method for scipy.optimize.minimize.
+
+        Returns
+        -------
+        array-like, 1d
+            The MAP predictor of the random effects.
+        """
+        x0 = self.prev_start
+
+        TOL = 1E-5
+        GTOL = 1E-5
+
+        d = self.n_groups * self.k_re
+        iters_allowed = 10000 * d
+        restarts_allowed = d
+
+        opts = {'gtol': GTOL, 'maxiter': iters_allowed}
+
+        for _ in range(1,restarts_allowed):
+
+            fun_grad_hess = self._gen_joint_like_grad_hess(fe_params, cov_re, scale, hess = False)
+            def fun(x): return fun_grad_hess(x)[0:2]
+
+            result = minimize(fun, x0, jac=True, tol=TOL, options= opts)
+
+            def func(x): return fun_grad_hess(x)[0]
+
+            if result.success:
+                break
+
+            x0 = np.random.normal(size = d)
+
+        if not result.success:
+            raise RuntimeError("Failed Laplace too many times.")
+
+        self.prev_start = result.x
+        mp = np.reshape(result.x, (self.n_groups, self.k_re))
+
+        return result.fun, mp
+
+    
+    def fit(self, start_params=None, niter_sa=0,
+            do_cg=True, fe_pen=None, cov_pen=None, free=None,
+            full_output=False, method='nm', **kwargs):
+        """
+        """
+
+        if start_params is None:
+            self.k_re2 = self.k_re * (self.k_re + 1) // 2
+            self.k_tot = self.k_fe + self.k_re2
+
+        run = 0
+        max_run = 10 * self.k_tot**2
+        keep_running = True
+
+        while keep_running:
+            try:
+                run += 1
+                start_params = np.random.uniform(size=self.k_tot)
+                rslt = super(MixedGLM, self).fit(start_params=start_params,
+                                                skip_hessian=True, method=method,
+                                                maxfun=10000, maxiter=10000,
+                                               **kwargs)
+                keep_running = False
+            except RuntimeError:
+                if run > max_run:
+                    raise
+
+        params = MixedGLMParams.from_packed(rslt.params, self.k_fe,
+                                            use_sqrt=self.use_sqrt)
+
+
+        # Convert to the final parameterization (i.e. undo the square
+        # root transform of the covariance matrix, and the profiling
+        # over the error variance).
+        fe_params = params.get_fe_params()
+        cov_re_unscaled = params.get_cov_re()
+        scale = self._scale
+        cov_re = scale * cov_re_unscaled
+
+        # TODO: Fix pcov
+        pcov = 1
```

Comment:
Yea, I don't really understand a lot of the stuff going on in fit. Not sure I follow your comment.